Support fzf and fzf.vim

noisysocks opened the pull request:
Adds a more "airliney" statusbar for the terminal buffers created by the [fzf](https://github.com/junegunn/fzf) and [fzf.vim](https://github.com/junegunn/fzf.vim) plugins.

| Command | Before | After |
| - | - | - |
| `:Files` | <img width="858" alt="Screen Shot 2020-06-05 at 17 00 46" src="https://user-images.githubusercontent.com/612155/83846690-585d0400-a74e-11ea-9e26-124469627472.png"> | <img width="858" alt="Screen Shot 2020-06-05 at 16 59 49" src="https://user-images.githubusercontent.com/612155/83846703-5d21b800-a74e-11ea-91e2-00ce44346508.png"> |
| `:Tags` | <img width="858" alt="Screen Shot 2020-06-05 at 17 00 52" src="https://user-images.githubusercontent.com/612155/83846731-690d7a00-a74e-11ea-9dee-7bf80667fb58.png"> | <img width="858" alt="Screen Shot 2020-06-05 at 16 59 58" src="https://user-images.githubusercontent.com/612155/83846756-7296e200-a74e-11ea-8284-e0d718dd1e61.png"> |

kazukazuinaina commented:
I used Vim and tried your patch, but it didn't behave the same as your screenshot.
Originally, Normal is displayed where FZF should be displayed.
Are you using Neovim?

## screen shot

<img width="562" alt="スクリーンショット 2020-06-07 0 11 51" src="https://user-images.githubusercontent.com/36619465/83947735-d9ea8a00-a853-11ea-8016-28aa6a873a83.png">
![Uploading スクリーンショット 2020-06-07 0.11.27.png…]()

noisysocks commented:
Yes, you're right, I'm using Neovim.

I spent some time looking into why this isn't working in Vim and there's two things going on.

- [The fzf plugin will only use a terminal buffer to run `fzf` when running Neovim, GVim, or when the user has specified a custom layout.](https://github.com/junegunn/fzf/blob/master/README-VIM.md#fzf-inside-terminal-buffer) This means that, in Vim, by default, `fzf` [runs using an `execute !`](https://github.com/junegunn/fzf/blob/master/plugin/fzf.vim#L597) and so won't have a statusline at all. I don't know if there's a way to force a statusline in this case. I am thinking that, in this case, it's best that vim-airline does nothing. Any thoughts?

- If you force Vim to use a terminal buffer by setting e.g. `let g:fzf_layout = { 'left': '50%' }`, a vim-airline statusline appears and `airline#extensions#fzf#apply` is invoked but `&filetype` is empty. It looks like `filetype` is set _after_ `airline#extensions#fzf#apply` is called. Is there a way to force airline to update the statusline when `filetype` is set? I tried adding `autocmd FileType fzf call airline#update_statusline()` but this did not work.

noisysocks commented:
_Oops, accidentally closed the PR._

I worked around the second issue I described in my comment above by setting the statusline manually when the filetype is set to fzf. It's the same approach [taken by the ctrlspace airline extension](https://github.com/vim-airline/vim-airline/blob/master/autoload/airline/extensions/ctrlspace.vim).

I also simplified the statusbar further so that only 'FZF' is displayed. Previously the current fzf.vim command and working directory were shown, but this is unnecessary since all of this is already shown in the prompt.

Looks like there are some lint errors. I'll try to fix them up tomorrow.

noisysocks commented:
I iterated on this a little more and was able to get the original approach using `add_statusline_func` working again. I'm disabling the custom statusline that [fzf.vim sets](https://github.com/junegunn/fzf.vim/blob/8f1e73b598d27d78dfb5843be19a73b6a3b222b1/plugin/fzf.vim#L118-L121) by running `autocmd!` in the fzf.vim augroup. This is hacky but I like it better than the other approach I tried which circumvents the vim-airline pipeline.

Again, there won't be any statusline when fzf is configured to not use a terminal buffer. I think this makes sense and it looks like the author of fzf is [considering moving to always using terminal buffers](https://github.com/junegunn/fzf/issues/1860#issuecomment-584522269) anyway.

I also added an inactive statusline and fixed all of the lint errors.

Here's an updated before and after in Vim:

| Before | After |
| - | - |
| <img width="1426" alt="Screen Shot 2020-06-13 at 10 43 54" src="https://user-images.githubusercontent.com/612155/84555919-da5ab780-ad62-11ea-93ab-bedc17b268d3.png"> | <img width="1426" alt="Screen Shot 2020-06-13 at 10 44 20" src="https://user-images.githubusercontent.com/612155/84555934-e6df1000-ad62-11ea-9eae-174bf688a1b6.png"> |

Overall, I'm pretty happy with this. It's a _very_ minor improvement but hopefully one that fzf+vim-airline users will appreciate and can build on in the future.

Let me know what you think.

chrisbra commented:
does it now work with vim?

kazukazuinaina commented:
It seems to work if I build this configuration.

```viml
let g:fzf_layout = { 'window': '10new' }
```

I don't think it works by default.

chrisbra commented:
okay let's merge it then. We can always improve later

igbanam commented:
I just saw FZF defaulted to the window format. Thought to bring this up here so we're aware of the changes this might need. https://github.com/junegunn/fzf/commit/c60ed1758315f0d993fbcbf04459944c87e19a48

chrisbra commented:
@igbanam would you mind creating a PR with the necessary changes?
